enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,546 @@
+snapshot of etcd and add it to the Ignition config for the host.
+After rebooting, we will use the restored `etcd-member` from the
+snapshot to rebuild the database. This allows etcd and the API service
+to come up on the host without having to re-apply all of the
+kubernetes operations run during bootstrapping.
+
+#### OpenShift-installer
+
+We will add a new `create single-node-ignition-config` command to the installer to create
+ `bootstrap-in-place-for-live-iso.ign` Ignition config.
+This new target will not output master.ign and worker.ign.
+Allow the user to specify the target disk drive for coreos-installer using environment
+ variable `OPENSHIFT_INSTALL_EXPERIMENTAL_BOOTSTRAP_IN_PLACE_COREOS_INSTALLER_ARGS`.
+
+This Ignition config will have a different `bootkube.sh` from the
+default bootstrap Ignition. In addition to the standard rendering
+logic, the modified script will:
+
+1. Start `cluster-bootstrap` without required pods by setting `--required-pods=''`
+2. Run `cluster-bootstrap` with the `--bootstrap-in-place` option.
+3. Fetch the master Ignition and combine it with the original Ignition
+   config, the control plane static pod manifests, the required
+   kubernetes resources, and the bootstrap etcd database snapshot to
+   create a new Ignition config for the host.
+3. Write the RHCOS image and the combined Ignition config to disk.
+4. Reboot the node.
+
+#### Cluster-bootstrap
+
+`cluster-bootstrap` will have a new entrypoint `--bootstrap-in-place`
+which will get the master Ignition as input and will enrich the master
+Ignition with control plane static pods manifests and all required
+resources, including the etcd database.
+
+`cluster-bootstrap` normally waits for a list of required pods to be
+ready. These pods are expected to start running on the control plane
+nodes when the bootstrap and control plane run in parallel. That is
+not possible when bootstrapping in place, so when `cluster-bootstrap`
+runs with the `--bootstrap-in-place` option it should only apply the
+manifests and then tear down the control plane.
+
+If `cluster-bootstrap` fails to apply some of the manifests, it should
+return an error.
+
+#### Bootstrap / Control plane static pods
+
+We will review the list of revisions for apiserver/etcd and see if we
+can reduce them by reducing revisions caused by observations of known
+conditions.  For example in a single node we know what the etcd
+endpoints will be in advance, We can avoid a revision by observing
+this post install.  This work will go a long way to reducing
+disruption during install and improve mean time to recovery for upgrade re-deployments
+and failures.
+
+The control plane components we will copy from
+`/etc/kubernetes/manifests` into the master Ignition are:
+
+1. etcd-pod
+2. kube-apiserver-pod
+3. kube-controller-manager-pod
+4. kube-scheduler-pod
+
+These components also require other files generated during bootstrapping:
+
+1. `/var/lib/etcd`
+2. `/etc/kubernetes/bootstrap-configs`
```

Comment:
Since this is copying bootstrap information into the final system (an explicit non-goal) what is the testing in place to be SURE that we never leak information past `post-reboot`?

Similar for /var/lib/etcd, though different. What do we have in place to make SURE that the /var/lib/etcd (or really the data IN that etcd) which turns up in the final master is exactly the same as I'd get from a non-bootstrapless setup? How are we testing and enforcing the 'no data crosses the boundary' goal of this enhancement?

Reply:
The goal is to get the final master is exactly the same as a non-bootstrapless setup.
We copying the bare minimum in order to complete the installation after the reboot.
Once the post-reboot service complete all the files that "belong" to the bootstrap get deleted. 
We should add a test that validates that.